enhance production guide with a few Argo specific guidelines [#20999][#28637] (#29078)
mistercrunch/Airflow: docs/helm-chart/index.rst

This is so these CI/CD services can perform updates without issues and preserve the immutability of Kubernetes Job manifests.

This also applies if you install the chart using ``--wait`` in your ``helm install`` command.

.. note::
    While deploying this Helm chart with Argo, you might encounter issues with database migrations not running automatically on upgrade.

> To ensure database migrations with Argo CD, you will need to add:

Reviewer: ```suggestion
To run database migrations with Argo CD automatically, you will need to add:
```

nit

> .. code-block::yaml

Reviewer 2: ```suggestion
.. code-block:: yaml
```

RST won't display this code block without a whitespace in between `::` and `yaml`.

    migrateDatabaseJob:
        jobAnnotations:
            "argocd.argoproj.io/hook": Sync

> This will run database migrations when the Airflow Docker image is upgraded. This approach has a limitation in that the database migrations will run every time there is a ``Sync`` event in Argo. This is a trade-off for automation at the cost of some computational loss.

Reviewer: ```suggestion
This will run database migrations every time there is a ``Sync`` event in Argo CD. While it is not ideal to run the migrations on every sync, it is a tradeoff that allows them to be run automatically.
```

Maybe something like this?

I think we could probably do without this, as this is the same situation with the helm hook. Though this might be my maintainer bias speaking here, so happy to leave it.

Author: I thought about it a bit, and I guess this is continuation of the previous section like you said, and we would be repeating that bit. Have applied your suggestions

> This will ensure database migrations run when the Airflow Docker image is upgraded. This approach has a limitation in that the database migrations will run every time there is a ``Sync`` event in Argo. This is a trade-off for automation at the cost of some computational loss.

Reviewer 2: ```suggestion
This will run database migrations when the Airflow Docker image is upgraded. This approach has a limitation in that the database migrations will run every time there is a ``Sync`` event in Argo. This is a trade-off for automation at the cost of some computational loss.
```

Avoid double usage of "ensure database migrations", reads a bit nicer.

> If you use the Celery(Kubernetes)Executor, and using the built-in Redis, it is recommended that you setup a static Redis password either by supplying ``redis.passwordSecretName`` and ``redis.data.brokerUrlSecretName`` or ``redis.password``.  See `Celery Backend <production-guide#celery-backend>`__ for more information about managing Celery Backend

Reviewer 2: ```suggestion
If you use the Celery(Kubernetes)Executor with the built-in Redis, it is recommended that you set up a static Redis password either by supplying ``redis.passwordSecretName`` and ``redis.data.brokerUrlSecretName`` or ``redis.password``.
```

Minor spelling nitpicks.

Also, would remove the last reference. That links to a section about managing your own Celery Backend, while the prior sentence speaks about managing the built-in Redis.

This is so these CI/CD services can perform updates without issues and preserve the immutability of Kubernetes Job manifests.

This also applies if you install the chart using ``--wait`` in your ``helm install`` command.

.. note::

>     While deploying this Helm chart with Argo, you might encounter issues with ``DB Migrations`` not running automatically on upgrade.

Reviewer 2: ```suggestion
    While deploying this Helm chart with Argo, you might encounter issues with database migrations not running automatically on upgrade.
```

Don't see a point in stating "DB migrations" as code-formatted text.

> To ensure DB Migrations with Argo CD, you will need to:

Reviewer 2: ```suggestion
To ensure database migrations with Argo CD, you will need to add:
```

Would simply write "DB" out fully for clarity. Also, "Migrations" is not a name so lowercase.

.. code-block::yaml

    migrateDatabaseJob:
        jobAnnotations:
            "argocd.argoproj.io/hook": Sync

> and, equivalent configurations in case of Flux CD

Reviewer 2: Is the code different for Flux CD? If so, could you provide an example?

Author: @BasPH tbh, I haven't used Flux CD, nor could I find anything on how to achieve this in Flux CD docs. Do you think its more appropriate to remove this line? I only put it in there because the original section mentioned Flux CD in the heading.

Author: My intent was to document:

- That there are 'specific' extra steps to make this helm chart work with Argo and Flux
- Specifically, resolve the documentation wrt to Argo because we have been getting a few issues mentioning helm-chart doesn't work with Argo

Reviewer: I'd suggest we drop this line if we don't know what we need for flux.

.. note::

>     This will ensure that DB Migrations run when Airflow Docker image is upgraded. This approach has a limitation that the Db Migrations will run every time there is a ``Sync`` event in Argo. This is a trade off for automation at the cost of some computational loss.

Reviewer 2: ```suggestion
    This will ensure database migrations run when the Airflow Docker image is upgraded. This approach has a limitation in that the database migrations will run every time there is a ``Sync`` event in Argo. This is a trade-off for automation at the cost of some computational loss.
```

> .. note::
>     This will ensure database migrations run when the Airflow Docker image is upgraded. This approach has a limitation in that the database migrations will run every time there is a ``Sync`` event in Argo. This is a trade-off for automation at the cost of some computational loss.

Reviewer: ```suggestion
This will ensure database migrations run when the Airflow Docker image is upgraded. This approach has a limitation in that the database migrations will run every time there is a ``Sync`` event in Argo. This is a trade-off for automation at the cost of some computational loss.
```

Probably doesn't need to be a note, as it's just a continuation of the stuff right above it.

Is this actually tied to image changes, or would any change kick it off (I'm not very familiar with Argo CD)?

Author: Other changes in the helm chart values for deployment will also kick it off, apart from the image changes. Which is why I added this a note as this does have some limitations.  I am happy to remove this as a note and convert it into normal text if you feel thats more appropriate

.. warning::

>     While deploying with Argo CD, you may encounter issues  with ``(built-in)Redis secrets`` (when you use Celery or CeleryKubernetesExecutor) not being rotated properly after Airflow pods restart. It's recommended that you bring your own Redis. See `Celery Backend <production-guide#celery-backend>`__

Reviewer 2: ```suggestion
    While deploying with Argo CD, you may encounter issues with (built-in) Redis secrets (when you use the Celery(Kubernetes)Executor) not being rotated properly after Airflow pods restart. It's recommended that you bring your own Redis. See `Celery Backend <production-guide#celery-backend>`__
```

Don't see the added benefit in code highlighting "(built-in)Redis secrets".

>     While deploying with Argo CD, you may encounter issues with (built-in) Redis secrets (when you use the Celery(Kubernetes)Executor) not being rotated properly after Airflow pods restart. It's recommended that you bring your own Redis. See `Celery Backend <production-guide#celery-backend>`__

Reviewer: And for the Argo CD case, instead of jumping straight to suggesting BYO redis, it might be better to suggest folks set `redis.passwordSecretName` and `data.brokerUrlSecretName`, or `redis.password` as a first option?

Author: I will make this change
